File: src/test/java/de/schramm/royalbash/controller/GameControllerTest.java
package de.schramm.royalbash.controller;

import de.schramm.royalbash.controller.service.core.Game;
import de.schramm.royalbash.controller.service.GameService;
import de.schramm.royalbash.controller.service.core.Player;
import de.schramm.royalbash.controller.service.core.State;
import de.schramm.royalbash.controller.service.gameevent.NoOpEvent;
import lombok.val;
import org.junit.Test;
import org.junit.runner.RunWith;
import org.mockito.Mockito;
import org.skyscreamer.jsonassert.JSONAssert;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.boot.test.autoconfigure.web.servlet.WebMvcTest;
import org.springframework.boot.test.mock.mockito.MockBean;
import org.springframework.http.MediaType;
import org.springframework.test.context.junit4.SpringRunner;
import org.springframework.test.web.servlet.MockMvc;
import org.springframework.test.web.servlet.request.MockMvcRequestBuilders;

import java.util.Optional;

import static org.assertj.core.api.Assertions.assertThat;
import static org.mockito.Matchers.any;
import static org.mockito.Mockito.times;
import static org.mockito.Mockito.verify;

@RunWith(SpringRunner.class)
@WebMvcTest(value = GameController.class, secure = false)
public class GameControllerTest {

    @Autowired
    private
    MockMvc mockMvc;

    @MockBean
    private GameService gameService;

    @Test
    public void should_deliver_game() throws Exception {

        // Given
        val game = Game.builder()
                .player1(Player.builder()
                        .name("Player 1")
                        .build())
                .player2(Player.builder().build())
                .playerOnTurn(Player.builder().build())
                .state(State.OPEN)
                .build();
        Mockito.when(gameService.retrieveGame(any())).thenReturn(Optional.of(game));

        val requestBuilder = MockMvcRequestBuilders
                .get("/game/1")
                .accept(MediaType.APPLICATION_JSON);

        // When
        val result = mockMvc.perform(requestBuilder)
                .andReturn()
                .getResponse()
                .getContentAsString();

        // Then
        JSONAssert.assertEquals("{\"player1\": {\"name\": \"Player 1\"}}", result, false);
    }

    @Test
    public void should_not_deliver_game_but_status_code_404_if_game_not_found() throws Exception {

        // Given
        Mockito.when(gameService.retrieveGame(any())).thenReturn(Optional.empty());

        val requestBuilder = MockMvcRequestBuilders
                .get("/game/1")
                .accept(MediaType.APPLICATION_JSON);

        // When
        val statusCode = mockMvc.perform(requestBuilder)
                .andReturn()
                .getResponse()
                .getStatus();

        // Then
        assertThat(statusCode).isEqualTo(404);
    }

    @Test
    public void should_deliver_new_game() throws Exception {

        // Given
        val accountId1 = "Account 1";
        val accountId2 = "Account 2";
        val game = Game.builder()
                .player1(Player.builder()
                        .name(accountId1)
                        .build())
                .player2(Player.builder()
                        .name(accountId2)
                        .build())
                .playerOnTurn(Player.builder().build())
                .state(State.OPEN)
                .build();
        Mockito.when(gameService.createGame(accountId1, accountId2)).thenReturn(game);

        val requestBuilder = MockMvcRequestBuilders
                .post("/game")
                .content(String.format(
                        "{\"accountId1\": \"%s\", \"accountId2\": \"%s\"}",
                        accountId1,
                        accountId2
                ))
                .accept(MediaType.APPLICATION_JSON)
                .contentType(MediaType.APPLICATION_JSON);

        // When
        val result = mockMvc.perform(requestBuilder)
                .andReturn()
                .getResponse()
                .getContentAsString();

        // Then
        JSONAssert.assertEquals(
                "{\"player1\": {\"name\": \"Account 1\"}, \"player2\": {\"name\": \"Account 2\"}}",
                result,
                false
        );
    }

    @Test
    public void should_resolve_event_and_return_updated_game() throws Exception {

        // Given
        val gameId = "Id 1";
        val game = Game.builder()
                .player1(Player.builder().build())
                .player2(Player.builder().build())
                .playerOnTurn(Player.builder().build())
                .state(State.OPEN)
                .build();
        Mockito.when(gameService.commitGameEvent(gameId, NoOpEvent.builder().build())).thenReturn(Optional.of(game));
        val requestBuilder = MockMvcRequestBuilders
                .post("/game/event")
                .content(String.format(
                        "{\"gameId\": \"%s\", \"event\": {\"type\": \"NO_OP\"}}",
                        gameId
                ))
                .accept(MediaType.APPLICATION_JSON)
                .contentType(MediaType.APPLICATION_JSON);

        // When
        val result = mockMvc.perform(requestBuilder)
                .andReturn()
                .getResponse()
                .getContentAsString();

        // Then
        verify(gameService, times(1)).commitGameEvent(gameId, NoOpEvent.builder().build());
        JSONAssert.assertEquals(
                "{}",
                result,
                false
        );
    }

    @Test
    public void should_not_resolve_event_but_return_status_code_404_if_game_not_found() throws Exception {

        // Given
        val gameId = "Id 1";
        Mockito.when(gameService.commitGameEvent(gameId, NoOpEvent.builder().build())).thenReturn(Optional.empty());
        val requestBuilder = MockMvcRequestBuilders
                .post("/game/event")
                .content(String.format(
                        "{\"gameId\": \"%s\", \"event\": {\"type\": \"NO_OP\"}}",
                        gameId
                ))
                .accept(MediaType.APPLICATION_JSON)
                .contentType(MediaType.APPLICATION_JSON);

        // When
        val statusCode = mockMvc.perform(requestBuilder)
                .andReturn()
                .getResponse()
                .getStatus();

        // Then
        assertThat(statusCode).isEqualTo(404);
    }
}

File: src/test/kotlin/de/schramm/royalbash/controller/GameControllerTest.kt
package de.schramm.royalbash.controller

import de.schramm.royalbash.controller.service.GameService
import de.schramm.royalbash.controller.service.core.Game
import de.schramm.royalbash.controller.service.core.Player
import de.schramm.royalbash.controller.service.core.State.OPEN
import de.schramm.royalbash.controller.service.gameevent.NoOpEvent
import io.mockk.every
import io.mockk.mockk
import io.mockk.verify
import org.assertj.core.api.Assertions.assertThat
import org.junit.Test
import org.junit.runner.RunWith
import org.skyscreamer.jsonassert.JSONAssert
import org.springframework.beans.factory.annotation.Autowired
import org.springframework.boot.test.autoconfigure.web.servlet.WebMvcTest
import org.springframework.boot.test.context.TestConfiguration
import org.springframework.context.annotation.Bean
import org.springframework.http.MediaType
import org.springframework.test.context.junit4.SpringRunner
import org.springframework.test.web.servlet.MockMvc
import org.springframework.test.web.servlet.request.MockMvcRequestBuilders
import java.util.*

@RunWith(SpringRunner::class)
@WebMvcTest(GameController::class)
class GameControllerTest {

    @TestConfiguration
    open class ControllerTestConfig {
        @Bean
        open fun gameService() = mockk<GameService>()
    }

    @Autowired
    private lateinit var mockMvc: MockMvc

    @Autowired
    private lateinit var gameService: GameService

    @Test
    @Throws(Exception::class)
    fun should_deliver_game() {

        // Given
        val game = Game(
                "1",
                player1 = Player("Id 2"),
                player2 = Player("Id 3"),
                playerOnTurn = Player("Id 3"),
                state = OPEN)
        every { gameService.retrieveGame("1") } returns Optional.of(game)

        val requestBuilder = MockMvcRequestBuilders
                .get("/game/1")
                .accept(MediaType.APPLICATION_JSON)

        // When
        val result = mockMvc.perform(requestBuilder)
                .andReturn()
                .response
                .contentAsString

        // Then
        JSONAssert.assertEquals("""{
                "id": "1",
                "player1": {"id": "Id 2"},
                "player2": {"id": "Id 3"},
                "playerOnTurn": "Id 3",
                "state": "OPEN"
            }""", result, false)
    }

    @Test
    @Throws(Exception::class)
    fun should_not_deliver_game_but_status_code_404_if_game_not_found() {

        // Given
        every { gameService.retrieveGame("1") } returns Optional.empty()

        val requestBuilder = MockMvcRequestBuilders
                .get("/game/1")
                .accept(MediaType.APPLICATION_JSON)

        // When
        val statusCode = mockMvc.perform(requestBuilder)
                .andReturn()
                .response
                .status

        // Then
        assertThat(statusCode).isEqualTo(404)
    }

    @Test
    @Throws(Exception::class)
    fun should_deliver_new_game() {

        // Given
        val accountId1 = "Account 1"
        val accountId2 = "Account 2"
        val game = Game(
                "Id 1",
                player1 = Player("Id 2", name = accountId1),
                player2 = Player("Id 3", name = accountId2),
                playerOnTurn = Player("Id 2"),
                state = OPEN)
        every { gameService.createGame(accountId1, accountId2) } returns game

        val requestBuilder = MockMvcRequestBuilders
                .post("/game")
                .content("""{
                    "accountId1": "Account 1",
                    "accountId2": "Account 2"
                }""")
                .accept(MediaType.APPLICATION_JSON)
                .contentType(MediaType.APPLICATION_JSON)

        // When
        val result = mockMvc.perform(requestBuilder)
                .andReturn()
                .response
                .contentAsString

        // Then
        JSONAssert.assertEquals(
                """{
                    "id": "Id 1",
                    "player1": {"id": "Id 2", "name": "Account 1"},
                    "player2": {"id": "Id 3", "name": "Account 2"},
                    "playerOnTurn": "Id 2",
                    "state": "OPEN"
                }""",
                result,
                false
        )
    }

    @Test
    @Throws(Exception::class)
    fun should_resolve_event_and_return_updated_game() {

        // Given
        val gameId = "1"
        val game = Game(
                gameId,
                player1 = Player("Id 2"),
                player2 = Player("Id 3"),
                playerOnTurn = Player("Id 2"),
                state = OPEN
        )
        every { gameService.commitGameEvent(gameId, NoOpEvent()) } returns Optional.of(game)

        val requestBuilder = MockMvcRequestBuilders
                .post("/game/1/event")
                .content("""{"event": {"type": "NO_OP"}}""")
                .accept(MediaType.APPLICATION_JSON)
                .contentType(MediaType.APPLICATION_JSON)

        // When
        val result = mockMvc.perform(requestBuilder)
                .andReturn()
                .response
                .contentAsString

        // Then
        verify { gameService.commitGameEvent(gameId, NoOpEvent()) }
        JSONAssert.assertEquals(
                "{}",
                result,
                false
        )
    }

    @Test
    @Throws(Exception::class)
    fun should_not_resolve_event_but_return_status_code_404_if_game_not_found() {

        // Given
        val gameId = "1"
        every { gameService.commitGameEvent(gameId, NoOpEvent()) } returns Optional.empty()
        val requestBuilder = MockMvcRequestBuilders
                .post("/game/1/event")
                .content("""{
                    "event": {"type": "NO_OP"}
                }""")
                .accept(MediaType.APPLICATION_JSON)
                .contentType(MediaType.APPLICATION_JSON)

        // When
        val statusCode = mockMvc.perform(requestBuilder)
                .andReturn()
                .response
                .status

        // Then
        assertThat(statusCode).isEqualTo(404)
    }
}